nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/SearchElasticsearchTest.java

```diff
@@ -0,0 +1,243 @@
+        runOnce(runner);
+        AbstractJsonQueryElasticsearchTest.testCounts(runner, 0, 1, 0, 0);
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("hit.count", "10");
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("page.number", "1");
+        assertState(runner.getStateManager(), paginationType, 10, 1);
+        runner.clearTransferState();
+
+        // second page
+        runOnce(runner);
+        AbstractJsonQueryElasticsearchTest.testCounts(runner, 0, 1, 0, 0);
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("hit.count", "10");
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("page.number", "2");
+        assertState(runner.getStateManager(), paginationType, 20, 2);
+        runner.clearTransferState();
+    }
+
+    @Override
+    public void testPagination(final PaginationType paginationType) throws Exception{
+        // test flowfile per page
+        final TestRunner runner = createRunner(false);
+        final TestElasticsearchClientService service = AbstractJsonQueryElasticsearchTest.getService(runner);
+        service.setMaxPages(2);
+        runner.setProperty(AbstractPaginatedJsonQueryElasticsearch.PAGINATION_TYPE, paginationType.getValue());
+        runner.setProperty(AbstractJsonQueryElasticsearch.QUERY, MATCH_ALL_WITH_SORT_BY_MSG_WITH_SIZE_QUERY);
+
+        // first page
+        runOnce(runner);
+        AbstractJsonQueryElasticsearchTest.testCounts(runner, 0, 1, 0, 0);
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("hit.count", "10");
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("page.number", "1");
+        assertState(runner.getStateManager(), paginationType, 10, 1);
+        runner.clearTransferState();
+
+        // second page
+        runOnce(runner);
+        AbstractJsonQueryElasticsearchTest.testCounts(runner, 0, 1, 0, 0);
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("hit.count", "10");
+        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("page.number", "2");
+        assertState(runner.getStateManager(), paginationType, 20, 2);
+        runner.clearTransferState();
+
+        // third page - no hits
+        runOnce(runner);
+        AbstractJsonQueryElasticsearchTest.testCounts(runner, 0, 0, 0, 0);
+        assertTrue(runner.getStateManager().getState(Scope.LOCAL).toMap().isEmpty());
+        reset(runner);
+
+
+        // test hits splitting
+        runner.setProperty(AbstractPaginatedJsonQueryElasticsearch.SEARCH_RESULTS_SPLIT, ResultOutputStrategy.PER_HIT.getValue());
```

Comment:
These tests could probably be refactored - but be careful to retain the `clearTransferState` use between the "pages" where `SEARCH_RESULTS_SPLIT` remains the same (e.g. `PER_RESPONSE` above, `PER_HIT` here, and `PER_QUERY` later)

Reply:
`testPagination` is not an actual unit test. It is an overridden method. It is called in the parent's `AbstractPaginatedJsonQueryElasticsearchTest` unit tests

1. `testScroll`
2. `testPit`
3. `testSearchAfter`

Reply:
You're right, I wonder whether these assertions could be re-written in a way similar to those in `PutElasticsearchJsonTest`, which also overrides the same method?

Reply:
I do not see a method named `testPagination` in `PutElasticsearchJsonTest`

Reply:
`SearchElssticsearch` (use your IDE's functionality to check for any other users of the method)

Reply:
`testPagination` is only implemented by `PaginatedJsonQueryElasticsearchTest` and `SearchElasticsearchTest` and it is called by `AbstractPaginatedJsonQueryElasticsearchTest` three times. Can you please elaborate on what you want? I do not fully understand the code and exactly how each of these test classes are exercising their features. Thanks!

Reply:
I added a `validatePagination` method which greatly simplified the `testPagination` method. Please let me know if that is what you had in mind.

Reply:
`SearchElasticsearchTest` uses `null` as one of the `ResultOutputStrategy` values, which defaults to `PER_RESPONSE` in `AbstractJsonQueryElasticsearch`, i.e. all `ResponseOutputStrategy` values are exercised in both test classes - this is probably something we should rationalise though, either use the `ResultOutputStrategy.values()` array as you've done in `PaginatedJsonQueryElasticsearchTest`, or create the array like in `SearchElasticsearchTest` with `null` - the danger with the latter is that it assumes the default will never change (which would be a breaking change, so likely wouldn't, but really it's a dangerous assumtion to have been made in the test code, so using the `.values()` approach is probably better really, so let's be consistent)

What you've done is certainly along the right lines though @dan-s1 - taking another look, I think most of the test setup & validation that's in `SearchElasticsearchTest` should probably be the same in `PaginatedJsonQueryElasticsearchTest` (where it's not already), how about:

* change the `AbstractPaginatedJsonQueryElasticsearchTest#testPagination` to **not** be `abstract` but instead provide the combined implementation from `SearchElasticsearchTest` & `PaginatedJsonQueryElasticsearchTest`
  * create runner, set pagination type (N.B. you've missed this completely from the new `PaginatedJsonQueryElasticsearchTest#testPagination` implementation), set query, iterate through all `ResultOutputStrategy.values()`
  * iterate over 3 pages in a loop, alernately setting `OUTPUT_NO_HITS` to `true` or `false` (or only a single iteration if it's ` PER_QUERY` test)
* call `validatePagination`, which looks like it should be the same/very similar to that already in `SearchElasticsearchTest`
* make `assertState` and `abstract` method, then implement in each test class
  * for `SearchElasticsearchTest`, this probably stays about the same, although we'll want to retain the `assertTrue(runner.getStateManager().getState(Scope.LOCAL).toMap().isEmpty());` entries where appropriate from `validatePagination` (and add in a call to `assertState` in their place)
  * for `PaginatedJsonQueryElasticsearchTest`, this is probably just always an `assertTrue(runner.getStateManager().assertStateNotSet();` check or such

Hope that makes sense

Reply:
Sort of. I pulled up `testPagination`, `validatePagination` and `assertState` as is from `SearchElasticsearchTest` to `AbstractPaginatedJsonQueryElasticsearchTest` and that works for both `SearchElasticsearchTest` and `PaginatedJsonQueryElasticsearchTest`. Not sure what you mean with the last bullet with making assertState abstract as the method in `SearchElasticsearchTest` is already being used in multiple methods. Will changing it adversely effect the other methods which use it?

Reply:
One second it did not work as I said, I forgot to remove `testPagination` from `PaginatedJsonQueryElasticsearchTest`. When I do then I have 3 failures for `testPit`, `testScroll` and `testSearchAfter`. They all seem to error in `testCounts` on
 `runner.assertTransferCount(AbstractJsonQueryElasticsearch.REL_ORIGINAL, original);`

where expected was 0 but actual was 1

Reply:
`validatePagination` from `SearchElasticsearchTest` as is will not work for `PaginatedJsonQueryElasticsearchTest`
as there are differences in the  call to `AbstractJsonQueryElasticsearchTest.testCounts` in the `validatePagination` of `SearchElasticsearchTest` and `PaginatedJsonQueryElasticsearchTest`